### PR TITLE
Makes to_array, from_array support custom numpy dtype, support float16 type in parser

### LINKED ISSUE #6152: Missing type support in parser for various types (float16, bfloat16, ...)
# Bug Report

### Is the issue related to model conversion?
No.

### Describe the bug

The parser has only partial support for data type parsing: https://github.com/onnx/onnx/blob/093a8d335a66ea136eb1f16b3a1ce6237ee353ab/onnx/defs/parser.cc#L436

The missing types are:
```
TensorProto_DataType_FLOAT16
TensorProto_DataType_BFLOAT16
TensorProto_DataType_FLOAT8E4M3FN
TensorProto_DataType_FLOAT8E4M3FNUZ
TensorProto_DataType_FLOAT8E5M2
TensorProto_DataType_FLOAT8E5M2FNUZ
TensorProto_DataType_COMPLEX64
TensorProto_DataType_COMPLEX128
```

### System information
System independent.

### Reproduction instructions
Find a sample model and instruction for float16 below; note however that it would be great to support all types not just float16. The error will be the same for the other types.

Sample model:
[gemmfloat16.onnxtxt](https://github.com/user-attachments/files/15513562/gemmfloat16.onnxtxt.txt)

Sample instructions:
```
import onnx
from pathlib import Path
onnx.parser.parse_model(Path("gemmfloat16.onnxtxt.txt").read_text())
```
Result:
`onnx.parser.ParseError: b'[ParseError at position (...)]\nError context: <float16[4, 4] weight = {...}, float16[4] bias = {...}>\nUnhandled type: %d10'`

(Note: The `.onnxtxt` format is not allowed to be uploaded therefore the sample is `.onnxtxt.txt`)
### Expected behavior
ONNX model is parsed successfully.

### PR DESCRIPTION
**Pull request by xadupre**
### Description
from_array, to_array do not support float 8 types, int4, uint4 bfloat16. It is confusing.
Fixes #6152 as well.

### CODE REVIEW

File: onnx/helper.py [hunk 1]
```diff
@@ -1599,9 +1598,38 @@ def np_dtype_to_tensor_dtype(np_dtype: np.dtype) -> int:
...
+    if np_dtype in mapping._NP_TYPE_TO_TENSOR_TYPE:
+        return cast(
+            int,
+            mapping._NP_TYPE_TO_TENSOR_TYPE[np_dtype],
+        )
+
+    if np.issubdtype(np_dtype, np.str_):
+        return TensorProto.STRING
+
+    if np_dtype == custom_np_types.float8e4m3fn and np_dtype.descr[0][0] == "e4m3fn":
```
Comment (gramalingam) [inline onnx/helper.py:1610]: Just trying to understand: why is the second check above required?

Comment (xadupre) [reply onnx/helper.py:1610]: I tried without and it failed. numpy does not seem to check the description, only the underlying type (uint8 here).

**Comment by justinchuby** [comment]
Do we want to make `custom_element_types.py` public? It would be helpful to understand the pros and cons of doing so. (And mention it in the PR description when we make a decision)

**Comment by xadupre** [comment]
> Do we want to make `custom_element_types.py` public? It would be helpful to understand the pros and cons of doing so. (And mention it in the PR description when we make a decision)

The types defined here are returned by to_array. Why not making them public?

File: onnx/reference/custom_element_types.py [hunk 2]
```diff
@@ -1,14 +1,20 @@
 # Copyright (c) ONNX Project Contributors
-#
+
 # SPDX-License-Identifier: Apache-2.0
 from __future__ import annotations
 
-import numpy as np
+import warnings
+
+from onnx.custom_element_types import bfloat16  # noqa: F401
```
Comment (justinchuby) [inline onnx/reference/custom_element_types.py:8]: It’s possible to define `__all__` and fix f401

File: onnx/reference/custom_element_types.py [hunk 3]
```diff
@@ -1,14 +1,20 @@
...
-bfloat16 = np.dtype((np.uint16, {"bfloat16": (np.uint16, 0)}))
-float8e4m3fn = np.dtype((np.uint8, {"e4m3fn": (np.uint8, 0)}))
-float8e4m3fnuz = np.dtype((np.uint8, {"e4m3fnuz": (np.uint8, 0)}))
-float8e5m2 = np.dtype((np.uint8, {"e5m2": (np.uint8, 0)}))
-float8e5m2fnuz = np.dtype((np.uint8, {"e5m2fnuz": (np.uint8, 0)}))
-uint4 = np.dtype((np.uint8, {"uint4": (np.uint8, 0)}))
-int4 = np.dtype((np.int8, {"int4": (np.int8, 0)}))
+warnings.warn(
+    "This file was moved to onnx.custom_element_types. Importing this module will fail in onnx>=1.19.0.",
+    category=DeprecationWarning,
```
Comment (justinchuby) [inline onnx/reference/custom_element_types.py:18]: Use future warning for deprecations. DeprecationWarnings are reserved for the Python language deprecations

File: onnx/test/parser_test.py [hunk 4]
```diff
@@ -268,6 +268,52 @@ def test_parse_various_float_values(self, test_literal, expect_exception):
...
+           >
+           agraph (float[N] X) => (__STYPE__[N] C)
+           <
+             __STYPE__[1] weight = {0}
+           >
+           {
+              C = Cast<to=__ITYPE__>(X)
+           }
+           """
+        text_graph = cast.replace("__ITYPE__", str(itype)).replace("__STYPE__", name)
```
Comment (justinchuby) [inline onnx/test/parser_test.py:311]: This can be done with str.format() and named arguments

**Review by justinchuby — APPROVED** [review]

File: onnx/custom_element_types.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
+# Copyright (c) ONNX Project Contributors
+#
+# SPDX-License-Identifier: Apache-2.0
+"""This module defines custom dtypes not supported by numpy.
```
Comment (gramalingam) [inline onnx/custom_element_types.py:4]: As discussed, I suggest adding the following disclaimer below:
```
The numpy representation dtypes used below are meant for internal use. They may change in the
future based on the industry standardization of these numpy types.
```

I wonder if we should name the file `_custom_element_types.py` perhaps?

File: onnx/reference/custom_element_types.py (hunk 2 above)
Comment (xadupre) [reply onnx/reference/custom_element_types.py:8]: done

File: onnx/reference/custom_element_types.py (hunk 3 above)
Comment (xadupre) [reply onnx/reference/custom_element_types.py:18]: done

File: onnx/test/parser_test.py (hunk 4 above)
Comment (xadupre) [reply onnx/test/parser_test.py:311]: done

File: onnx/custom_element_types.py (hunk 5 above)
Comment (xadupre) [reply onnx/custom_element_types.py:4]: It is not really internal, it is used by the reference implementation as well.

Comment (justinchuby) [reply onnx/custom_element_types.py:4]: `reference implementation` = internal so I think it's ok. If we are talking about the output values by the reference implementation which takes these types, then I suggest we convey a message of "don't care what the type is except for knowing what the byte representation is."

File: onnx/reference/custom_element_types.py [hunk 6]
```diff
@@ -5,16 +5,27 @@
...
 
 warnings.warn(
     "This file was moved to onnx.custom_element_types. Importing this module will fail in onnx>=1.19.0.",
-    category=DeprecationWarning,
+    category=PendingDeprecationWarning,
     stacklevel=1,
 )
+
+
+__all__ = [
```
Comment (justinchuby) [inline onnx/reference/custom_element_types.py:23]: nit: All should be placed before the imports: https://peps.python.org/pep-0008/#module-level-dunder-names

> Module level “dunders” (i.e. names with two leading and two trailing underscores) such as `__all__`, `__author__`, `__version__`, etc. should be placed after the module docstring but before any import statements except from `__future__` imports. Python mandates that future-imports must appear in the module before any other code except docstrings:

File: onnx/numpy_helper.py [hunk 7]
```diff
@@ -397,6 +470,62 @@ def from_array(arr: np.ndarray, name: str | None = None) -> TensorProto:
...
+        dt_to = np.int8  # type: ignore[assignment]
+    elif dt == custom_np_types.uint4 and dt.descr[0][0] == "uint4":
+        to = TensorProto.UINT4
+        dt_to = np.uint8  # type: ignore[assignment]
+    else:
+        return _from_array(tensor, name)
+
+    if to in (TensorProto.UINT4, TensorProto.INT4):
+        value = tensor.astype(dt_to).ravel()
+        buffer = (value[::2] & 0x0F) + (value[1::2] << 4)
```
Comment (justinchuby) [inline onnx/numpy_helper.py:514]: Do we need to handle odd-sized arrays? https://github.com/microsoft/onnxscript/blob/1aa7a7017225f95621e9a2008041056661ff7977/onnxscript/ir/_type_casting.py#L23-L25

File: onnx/numpy_helper.py [hunk 8]
```diff
@@ -8,6 +8,7 @@
 
 import numpy as np
 
+import onnx.custom_element_types as custom_np_types
```
Comment (justinchuby) [inline onnx/numpy_helper.py:11]: Curious if there was a need to rename? If not:

```suggestion
from onnx import custom_element_types
```

File: onnx/helper.py [hunk 9]
```diff
@@ -23,6 +23,7 @@
 import google.protobuf.message
 import numpy as np
 
+import onnx.custom_element_types as custom_np_types
```
Comment (justinchuby) [inline onnx/helper.py:26]: ditto

File: onnx/reference/custom_element_types.py (hunk 6 above)
Comment (xadupre) [reply onnx/reference/custom_element_types.py]: removed

File: onnx/numpy_helper.py (hunk 7 above)
Comment (xadupre) [reply onnx/numpy_helper.py:514]: We probably should.

File: onnx/numpy_helper.py (hunk 8 above)
Comment (xadupre) [reply onnx/numpy_helper.py]: I let this syntax since I rename the module with an underscore.

File: onnx/helper.py (hunk 9 above)
Comment (xadupre) [reply onnx/helper.py]: same

File: onnx/numpy_helper.py (hunk 8 above)
Comment (xadupre) [reply onnx/numpy_helper.py]: Having lines too long means I struggle with the linter.

File: docs/docsgen/source/api/custom_element_types.md
Comment (justinchuby) [inline docs/docsgen/source/api/custom_element_types.md:1]: We may not need to document this now that it is internal.

File: onnx/reference/ops/op_batch_normalization.py [hunk 10]
```diff
@@ -51,7 +51,18 @@ def _batchnorm_training_mode(
...
+        self,
+        x,
+        scale,
+        bias,
+        mean,
+        var,
+        epsilon=None,
+        is_test=None,
+        momentum=None,
+        spatial=None,
```
Comment (justinchuby) [inline onnx/reference/ops/op_batch_normalization.py:64]:
```suggestion
        spatial=None,  # noqa: ARG002
```

File: onnx/reference/ops/op_batch_normalization.py [hunk 11]
```diff
@@ -51,7 +51,18 @@ def _batchnorm_training_mode(
...
+    ):  # type: ignore  # noqa: ARG002
```
Comment (justinchuby) [inline onnx/reference/ops/op_batch_normalization.py:65]:
```suggestion
    ):  # type: ignore
```

File: onnx/test/helper_test.py [hunk 12]
```diff
@@ -1022,6 +1031,18 @@ def test_attr_type_to_str_undefined(self):
...
         self.assertEqual(result, "UNDEFINED")
 
+    def test_custom_types(self):
+        def _get(name):
+            if hasattr(_custom_element_types, name):
+                return getattr(_custom_element_types, name)
+            name = f"float8{name}"
+            if hasattr(_custom_element_types, name):
+                return getattr(_custom_element_types, name)
+            raise AttributeError(f"Unknown name {name!r}")
```
Comment (justinchuby) [inline onnx/test/helper_test.py:1041]:
```suggestion
            getattr(_custom_element_types, name)
```

**Review by justinchuby — APPROVED** [review]

File: onnx/numpy_helper.py (hunk 7 above)
Comment (xadupre) [reply onnx/numpy_helper.py:519]: I raised an exception. Odd size should not happen.

File: docs/docsgen/source/api/custom_element_types.md
Comment (xadupre) [reply docs/docsgen/source/api/custom_element_types.md:1]: Users will see it eventually. More documentation does not hurt I think.

**Review by gramalingam — APPROVED** [review]

**Pull request merged**